{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Long lista ficheros:  5\n",
      "Tiempo:  10/03/2021, 23:17:10\n",
      "0 ['./SecsGRANDES\\\\BRCA1_human_gene.fasta']\n",
      "./SecsGRANDES\\BRCA1_human_gene.fasta\n",
      "SequenceID = NG_005905.2\n",
      "Description = Homo sapiens BRCA1 DNA repair associated (BRCA1), RefSeqGene (LRG_292) on chromosome 17\n",
      "\n",
      "Tiempo:  10/04/2021, 07:00:12\n",
      "1 ['./SecsGRANDES\\\\EGFR_Human.fasta']\n",
      "./SecsGRANDES\\EGFR_Human.fasta\n",
      "SequenceID = NG_007726.3\n",
      "Description = Homo sapiens epidermal growth factor receptor (EGFR), RefSeqGene (LRG_304) on chromosome 7\n",
      "\n",
      "Tiempo:  10/04/2021, 19:06:03\n",
      "2 ['./SecsGRANDES\\\\ESR1_human.fasta']\n",
      "./SecsGRANDES\\ESR1_human.fasta\n",
      "SequenceID = NG_008493.2\n",
      "Description = Homo sapiens estrogen receptor 1 (ESR1), RefSeqGene (LRG_992) on chromosome 6\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Programa para calcular la correlación de una serie temporal con el método de DFA. La serie temporal ha sido transformada a\n",
    "## partir de una secuencia de nucleotidos siguiendo una regla de mapeo binaria. Para formar esta serie temporal tenemos que \n",
    "## recorrer la secuencia y si encontramos una Adenina(A), una Timina(T) o Uracilo(U) el valor de la serie temporal crece y si \n",
    "## encontramos una Citosina (C) o una Guanina (G) el valor de la serie decrece. Es una forma distinta de trasnformar las secuencias \n",
    "## de nucleotidos que la que propuso Peng et al. en 1992 [Long-range correlations in nucleotide sequences. Nature.], basada en la complementaridad de estos nucleotidos en las cadenas \n",
    "## de doble helice, en vez de la ocmposición química de los compuestos.\n",
    "\n",
    "##PAQUETES\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "# from Bio.SeqRecord import SeqRecord\n",
    "# import time\n",
    "# from scipy.optimize import curve_fit\n",
    "# from graph_tools import Graph\n",
    "# import random\n",
    "from scipy import stats\n",
    "# import math\n",
    "# import scipy.optimize as spo\n",
    "import datetime\n",
    "\n",
    "\n",
    "## Transformar el formato .fasta (formato en el que se descargan las secuencias del NCBI) a una string con la secuencia de nucleotidos\n",
    "\n",
    "def FastaToSeq(DNA_file):\n",
    "    file_in =DNA_file\n",
    "\n",
    "    #file_out='gene_seq_out.fasta'\n",
    "\n",
    "    fin = open(file_in,mode='r')\n",
    "    for record in SeqIO.parse(fin, 'fasta'):\n",
    "        rID = record.id\n",
    "        rDESC = record.description\n",
    "        rSEQ = record.seq\n",
    "    rDESC_2 = ' '.join(rDESC.split()[1:])\n",
    "    print('SequenceID = '  + rID)\n",
    "    print('Description = ' + rDESC_2 + '\\n')\n",
    "    # write new fasta file\n",
    "    #r=SeqIO.write(seq_record, f_out, 'fasta')\n",
    "    #if r!=1: print('Error while writing sequence:  ' + seq_record.id)\n",
    "    DNA_seq= rSEQ\n",
    "    DESC = rDESC_2\n",
    "    return DNA_seq,DESC,rID\n",
    "\n",
    "\n",
    "##Ubicación de la carpeta con las secuencias a analizar\n",
    "\n",
    "folder = './SecsGRANDES'\n",
    "\n",
    "## Dataframe donde almacenar los datos \n",
    "\n",
    "columnas1 = [\"Identificador\",\n",
    "            'NombreSecuencia',\n",
    "           'LongitudSecuencia',\n",
    "            'PENG-alpha',\n",
    "            'PENG-R2'\n",
    "           ]\n",
    "\n",
    "dtt_Peng = pd.DataFrame(columns=columnas1)\n",
    "\n",
    "## Lista con los ficheros .fasta que vamos a analizar \n",
    "listaFicheros=[]\n",
    "for root, dirs, files in os.walk(folder):\n",
    "    for file in files:\n",
    "        #print(root,dirs,file)\n",
    "        if file.endswith('.fasta') :\n",
    "            listaFicheros.append([os.path.join(root,file)])\n",
    "            \n",
    "print('Long lista ficheros: ',len(listaFicheros))\n",
    "\n",
    "#R2 que queremos para nuestro análisis\n",
    "R2paraPeng=0.995\n",
    "\n",
    "## Creamos el DNA WALK y  calculamos su correlación \n",
    "for iii,ff in enumerate(listaFicheros):\n",
    "\n",
    "    now = datetime.datetime.now()\n",
    "\n",
    "    print('Tiempo: ',now.strftime(\"%m/%d/%Y, %H:%M:%S\"))\n",
    "    print(iii,ff)\n",
    "    print(ff[0])\n",
    "\n",
    "\n",
    "    DNA_seq,dscrp,rID = FastaToSeq(ff[0])\n",
    "    DNA_seq= \"O\" + DNA_seq\n",
    "    length_seq=len(DNA_seq)\n",
    "    ##DNA WALK\n",
    "    DNA_walk=[]\n",
    "    x=np.arange(length_seq)\n",
    "    y=0\n",
    "    for i in range(length_seq):\n",
    "        if DNA_seq[i] in [\"C\",\"G\"]:\n",
    "            y=y -1\n",
    "        elif DNA_seq[i] in [\"T\",\"A\",\"U\"]:\n",
    "            y=y +1\n",
    "        else:\n",
    "            y=y\n",
    "\n",
    "        DNA_walk.append(y)\n",
    "        \n",
    "    F=[]\n",
    "    L=[] #log de L\n",
    "    # valores permisibles de l, serán:\n",
    "    lposibles = np.arange(1,int(np.floor(length_seq/1.5)+1))\n",
    "    #calculamos la correlación \n",
    "    # F^2(l) \n",
    "    for l in lposibles:\n",
    "        fluct2 = 0\n",
    "        Delta_y = []\n",
    "        Delta_y_2 = []\n",
    "        for l0 in range(0,length_seq-l):\n",
    "            Delta_y.append((DNA_walk[l0+l]-DNA_walk[l0]))\n",
    "\n",
    "            Delta_y_2.append((DNA_walk[l0+l]- DNA_walk[l0])**2)\n",
    "\n",
    "        Delta_y_mean = np.mean(Delta_y)\n",
    "\n",
    "        Delta_y_2_mean = np.mean(Delta_y_2)\n",
    "\n",
    "\n",
    "        fluct2 = Delta_y_2_mean-Delta_y_mean**2\n",
    "\n",
    "\n",
    "        ff = np.sqrt(fluct2)\n",
    "        if ff==0:\n",
    "            F.append(0)\n",
    "        else:\n",
    "            F.append(np.log10(ff))\n",
    "\n",
    "\n",
    "    L = np.log10(lposibles)\n",
    "\n",
    "\n",
    "    slope_list=[]\n",
    "    R2=[]\n",
    "    Lmaxposibles=[]\n",
    "    Nposibles=np.arange(30,1,-0.5)\n",
    "    for i in Nposibles:\n",
    "        Lmax=int(np.floor(length_seq/i))\n",
    "        Lsub=L[0:Lmax]\n",
    "        Fsub=F[0:Lmax]\n",
    "        slope, intercept, r_value, p_value, std_err = stats.linregress(Lsub,Fsub)\n",
    "        R2_value=r_value**2\n",
    "        Lmaxposibles.append(Lmax)\n",
    "\n",
    "\n",
    "        slope_list.append(slope)\n",
    "        R2.append(R2_value)\n",
    "        \n",
    "\n",
    "    #Seleccionamos el resultado con el valor de R2 que queremos obtener\n",
    "    indice = np.argmin(np.abs(np.asarray(R2)-R2paraPeng))\n",
    "    R2obtenido = R2[indice]\n",
    "    Noptimo = Nposibles[indice]\n",
    "    SlopeOptimo = slope_list[indice]\n",
    "    \n",
    "    \n",
    "    dtt_Peng=dtt_Peng.append({'Identificador':rID,'NombreSecuencia':dscrp,'LongitudSecuencia':length_seq,'PENG-alpha':SlopeOptimo,\"PENG-R2\":R2obtenido},ignore_index=True)\n",
    "#Guardamos los resultados\n",
    "print(dtt_Peng)\n",
    "dtt_Peng.to_csv(path_or_buf=\"Resultados_correlacion_DNAWALKS.csv\", sep=\";\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
